### ISSUE TITLE
prioritise replication of endangered chunks

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#244]
In 2.5.4 replication of "undergoal" chunks is done in parallel with replication of "endangered" chunks. I suggest to prioritise replication of "endangered" chunks (only one replica) over "undergoal" chunks if latter already have two or more replicas.

The idea is to replicate chunks with least number of replicas first, before replicating those that are less endangered.

**Comment by 4Dolio** [comment]
I very very much support this request and recommend that it be given the highest priority!...  ;)
This was one of my most desired changes which I noticed many years ago...

**Comment by psarna** [comment]
I'm preparing a solution to this problem: http://cr.skytechnology.pl:8081/#/c/1474/5 ,
but it may take a while until it's merged

**Comment by psarna** [comment]
This commit is in final testing stage now, please let me know if you have any remarks

**Comment by onlyjob** [comment]
Although it builds successfully on x86_64, it FTBFS on x86_32 (Debian i386):

~~~~
[ 44%] Building CXX object src/master/CMakeFiles/master.dir/chunks.cc.o
/var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc: In function 'void chunk_jobs_main()':
/var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:2200:93: error: no matching function for call to 'min(uint64_t&, std::deque<chunk*>::size_type)'
   endangeredToServe = std::min(gEndangeredChunksServingLimit, chunk::endangeredChunks.size());
                                                                                             ^
/var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:2200:93: note: candidates are:
In file included from /usr/include/c++/4.9/bits/char_traits.h:39:0,
                 from /usr/include/c++/4.9/string:40,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/media_label.h:4,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/chunk_with_address_and_label.h:5,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.h:26,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:20:
/usr/include/c++/4.9/bits/stl_algobase.h:194:5: note: template<class _Tp> const _Tp& std::min(const _Tp&, const _Tp&)
     min(const _Tp& __a, const _Tp& __b)
     ^
/usr/include/c++/4.9/bits/stl_algobase.h:194:5: note:   template argument deduction/substitution failed:
/var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:2200:93: note:   deduced conflicting types for parameter 'const _Tp' ('long long unsigned int' and 'std::deque<chunk*>::size_type {aka unsigned int}')
   endangeredToServe = std::min(gEndangeredChunksServingLimit, chunk::endangeredChunks.size());
                                                                                             ^
In file included from /usr/include/c++/4.9/bits/char_traits.h:39:0,
                 from /usr/include/c++/4.9/string:40,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/media_label.h:4,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/chunk_with_address_and_label.h:5,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.h:26,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:20:
/usr/include/c++/4.9/bits/stl_algobase.h:240:5: note: template<class _Tp, class _Compare> const _Tp& std::min(const _Tp&, const _Tp&, _Compare)
     min(const _Tp& __a, const _Tp& __b, _Compare __comp)
     ^
/usr/include/c++/4.9/bits/stl_algobase.h:240:5: note:   template argument deduction/substitution failed:
/var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:2200:93: note:   deduced conflicting types for parameter 'const _Tp' ('long long unsigned int' and 'std::deque<chunk*>::size_type {aka unsigned int}')
   endangeredToServe = std::min(gEndangeredChunksServingLimit, chunk::endangeredChunks.size());
                                                                                             ^
In file included from /usr/include/c++/4.9/algorithm:62:0,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/goal.h:4,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/chunks_availability_state.h:5,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.h:27,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:20:
/usr/include/c++/4.9/bits/stl_algo.h:3439:5: note: template<class _Tp> _Tp std::min(std::initializer_list<_Tp>)
     min(initializer_list<_Tp> __l)
     ^
/usr/include/c++/4.9/bits/stl_algo.h:3439:5: note:   template argument deduction/substitution failed:
/var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:2200:93: note:   mismatched types 'std::initializer_list<_Tp>' and 'long long unsigned int'
   endangeredToServe = std::min(gEndangeredChunksServingLimit, chunk::endangeredChunks.size());
                                                                                             ^
In file included from /usr/include/c++/4.9/algorithm:62:0,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/goal.h:4,
                 from /var/tmp/src/lfs_i/lizardfs/src/common/chunks_availability_state.h:5,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.h:27,
                 from /var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:20:
/usr/include/c++/4.9/bits/stl_algo.h:3444:5: note: template<class _Tp, class _Compare> _Tp std::min(std::initializer_list<_Tp>, _Compare)
     min(initializer_list<_Tp> __l, _Compare __comp)
     ^
/usr/include/c++/4.9/bits/stl_algo.h:3444:5: note:   template argument deduction/substitution failed:
/var/tmp/src/lfs_i/lizardfs/src/master/chunks.cc:2200:93: note:   mismatched types 'std::initializer_list<_Tp>' and 'long long unsigned int'
   endangeredToServe = std::min(gEndangeredChunksServingLimit, chunk::endangeredChunks.size());
                                                                                             ^
src/master/CMakeFiles/master.dir/build.make:123: recipe for target 'src/master/CMakeFiles/master.dir/chunks.cc.o' failed
make[4]: *** [src/master/CMakeFiles/master.dir/chunks.cc.o] Error 1
~~~~

**Comment by psarna** [comment]
Fixed, thanks

**Comment by onlyjob** [comment]
Thank you, it builds well. I'll test. Meanwhile maybe you could fix similar build problem #197 please? ;)
